core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala

```diff
@@ -52,43 +52,50 @@ class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: Transa
 
       val writeTxnMarkerResponse = response.responseBody.asInstanceOf[WriteTxnMarkersResponse]
 
-      for (txnMarker: TxnMarkerEntry <- txnMarkerEntries) {
-        val errors = writeTxnMarkerResponse.errors(txnMarker.producerId())
+      for (txnIdAndMarker: TxnIdAndMarkerEntry <- txnIdAndMarkerEntries) {
+        val transactionalId = txnIdAndMarker.txnId
+        val txnMarker = txnIdAndMarker.txnMarkerEntry
 
-        if (errors == null)
-          throw new IllegalStateException("WriteTxnMarkerResponse does not contain expected error map for pid " + txnMarker.producerId())
+        txnStateManager.getTransactionState(transactionalId) match {
+          case None =>
+            // txn topic partition has likely emigrated, just cancel it from the purgatory
+            txnMarkerPurgatory.cancelForKey(transactionalId)
 
-        val retryPartitions: mutable.Set[TopicPartition] = mutable.Set.empty[TopicPartition]
-        for ((topicPartition: TopicPartition, error: Errors) <- errors) {
-          error match {
-            case Errors.NONE =>
-              transactionMarkerChannel.pendingTxnMetadata(txnTopicPartition, txnMarker.producerId()) match {
-                case None =>
-                  // TODO: probably need to respond with Errors.NOT_COORDINATOR
-                  throw new IllegalArgumentException(s"transaction metadata not found during write txn marker request. partition ${txnTopicPartition} has likely emigrated")
-                case Some(metadata) =>
-                  // do not synchronize on this metadata since it will only be accessed by the sender thread
-                  metadata.topicPartitions -= topicPartition
+          case Some(epochAndMetadata) =>
+            val retryPartitions: mutable.Set[TopicPartition] = mutable.Set.empty[TopicPartition]
+            val errors = writeTxnMarkerResponse.errors(txnMarker.producerId)
+
+            if (errors == null)
+              throw new IllegalStateException(s"WriteTxnMarkerResponse does not contain expected error map for pid ${txnMarker.producerId}")
+
+            epochAndMetadata synchronized {
+              for ((topicPartition: TopicPartition, error: Errors) <- errors) {
+                error match {
+                  case Errors.NONE =>
+                    epochAndMetadata.transactionMetadata.topicPartitions -= topicPartition
```

Comment:
Structure wise, maybe it's better to do this as part of TransactionMetadata.completeTransitionTo() so that we can limit the place where internals of TransactionMetadata are modified?